### PR TITLE
feat(code_mappings): Order of stackframes not to affect results

### PR DESCRIPTION
**Pull request by armenzg**
This guarantees the same result with the same trees + stackframes input.

It also gets rid of `Any` usage for data typing.

Fixes WOR-2317

### CODE REVIEW

File: tests/sentry/integrations/github/test_integration.py [hunk 1]
```diff
@@ -629,5 +628,5 @@ def test_derive_code_mappings(self):
             "sentry/identity/oauth2.py",
         ]
         code_mappings = derive_code_mappings(stacktraces, trees)
-        # Order matters, this is why we only derive one of the two code mappings
-        assert code_mappings == [expected_code_mappings[0]]
+        # The reprocess feature allows determinging the sentry_plugins code mappings
+        assert code_mappings == expected_code_mappings
```
Comment (armenzg) [inline tests/sentry/integrations/github/test_integration.py:632]: This is the feature this PR gives us.

File: src/sentry/integrations/utils/tree.py [hunk 2]
```diff
@@ -1,9 +1,7 @@
 from typing import List
 
-from sentry.utils.json import JSONData
 
-
-def trim_tree(tree: JSONData, languages: List[str]) -> List[str]:
+def trim_tree(tree: List[str], languages: List[str]) -> List[str]:
```
Comment (armenzg) [inline src/sentry/integrations/utils/tree.py:4]: Removing the use of `Any` (`JSONData` is `Any`)

File: src/sentry/integrations/utils/code_mapping.py [hunk 3]
```diff
@@ -139,24 +149,21 @@ def _generate_code_mapping_from_tree(
         )
 
     def _matches_current_code_mappings(self, src_file: str, frame_filename: FrameFilename) -> bool:
-        if not self.reverse_mapping:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:142]: Removed this condition.

File: src/sentry/integrations/utils/code_mapping.py [hunk 4]
```diff
@@ -139,24 +149,21 @@ def _generate_code_mapping_from_tree(
...
-            # that source path.
-            #
-            # For instance sentry_plugins/slack/client.py matches these files
-            # - "src/sentry_plugins/slack/client.py",
-            # - "src/sentry/integrations/slack/client.py",
-            return any(
-                source_path
-                for source_path in self.reverse_mapping
-                if src_file.startswith(f"{source_path}/")
-            )
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:155]: Left shifted this block.

File: src/sentry/integrations/utils/code_mapping.py [hunk 5]
```diff
@@ -139,24 +149,21 @@ def _generate_code_mapping_from_tree(
...
+        # that source path.
+        #
+        # For instance sentry_plugins/slack/client.py matches these files
+        # - "src/sentry_plugins/slack/client.py",
+        # - "src/sentry/integrations/slack/client.py",
+        return any(
+            code_mapping.source_path
+            for code_mapping in self.code_mappings.values()
+            if src_file.startswith(f"{code_mapping.source_path}/")
+        )
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:162]: Using now `self.code_mappings` in the way I was using `self.reverse_mapping`.

File: src/sentry/integrations/utils/code_mapping.py [hunk 6]
```diff
@@ -73,23 +71,35 @@ def stacktrace_buckets(self, stacktraces: List[str]) -> Dict[str, Any]:
...
+        buckets: Dict[str, List[FrameFilename]] = self.stacktrace_buckets(stacktraces)
+
+        # We reprocess stackframes until we are told that no code mappings were produced
+        # This is order to reprocess past stackframes in light of newly discovered code mappings
+        # This allows for idempotency since the order of the stackframes will not matter
+        # This has no performance issue because stackframes that match an existing code mapping
+        # will be skipped
+        while True:
+            if not self.process_stackframes(buckets):
+                break
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:102]: If we introduce a bug in the call, this loop will make the tests run infinitely, thus, we're safe using `while True`.

File: src/sentry/integrations/utils/code_mapping.py [hunk 7]
```diff
@@ -73,23 +71,35 @@ def stacktrace_buckets(self, stacktraces: List[str]) -> Dict[str, Any]:
                 continue
         return buckets
 
-    def generate_code_mappings(self, stacktraces: List[str]) -> List[CodeMapping]:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:76]: If you make your browser window wide and use the Split view, the diff will be easier to view:
image

File: src/sentry/integrations/utils/code_mapping.py [hunk 8]
```diff
@@ -73,23 +71,35 @@ def stacktrace_buckets(self, stacktraces: List[str]) -> Dict[str, Any]:
...
-        """Generate code mappings based on the initial trees object and the list of stack traces"""
-        # We need to make sure that calling this method with a new list of stack traces
-        # should always start with a clean slate
-        self.code_mappings = {}
-        buckets: Dict[str, Any] = self.stacktrace_buckets(stacktraces)
-
+    def process_stackframes(self, buckets: Dict[str, List[FrameFilename]]) -> bool:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:74]: Typing improvement: `List[str]` -> `List[FrameFilename]`

File: src/sentry/integrations/utils/code_mapping.py [hunk 9]
```diff
@@ -48,14 +48,12 @@ def __repr__(self) -> str:
...
         self.trees = trees
-        self.code_mappings: Dict[str, Any] = {}
-        # This simplifies excluding source files for existing code mappings
-        self.reverse_mapping: List[str] = []
+        self.code_mappings: Dict[str, CodeMapping] = {}
 
-    def stacktrace_buckets(self, stacktraces: List[str]) -> Dict[str, Any]:
-        buckets: Dict[str, Any] = {}
+    def stacktrace_buckets(self, stacktraces: List[str]) -> Dict[str, List[FrameFilename]]:
+        buckets: Dict[str, List[FrameFilename]] = {}
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:56]: Typing improvement.

File: src/sentry/integrations/utils/code_mapping.py [hunk 10]
```diff
@@ -48,14 +48,12 @@ def __repr__(self) -> str:
 
 
 class CodeMappingTreesHelper:
-    def __init__(self, trees: Dict[str, Any]):
+    def __init__(self, trees: Dict[str, List[str]]):
         self.trees = trees
-        self.code_mappings: Dict[str, Any] = {}
-        # This simplifies excluding source files for existing code mappings
-        self.reverse_mapping: List[str] = []
+        self.code_mappings: Dict[str, CodeMapping] = {}
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:53]: Typing improvement.

File: src/sentry/integrations/utils/code_mapping.py [hunk 11]
```diff
@@ -15,7 +15,7 @@ class CodeMapping(NamedTuple):
     source_path: str
 
 
-def derive_code_mappings(stacktraces: List[str], trees: Dict[str, Any]) -> List[CodeMapping]:
+def derive_code_mappings(stacktraces: List[str], trees: Dict[str, List[str]]) -> List[CodeMapping]:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:18]: Typing improvement.

**Review by snigdhas — APPROVED** [review]

**Pull request merged**